### PR TITLE
[WIP] Remove Lock for Type in PyObject and impl Heap PyType Structure

### PR DESCRIPTION
**Pull request by qingshi163**
Do you agree with this appoarch?
@youknowone

### CODE REVIEW

File: vm/src/builtins/type.rs [hunk 1]
```diff
@@ -32,6 +34,28 @@ pub struct PyType {
...
 
+#[repr(C)]
+pub struct PyHeapType {
+    pub typ: PyType,
+    pub number_methods: PyRwLock<PyNumberMethods>,
+}
+
+impl From<PyHeapType> for PyType {
+    fn from(heaptype: PyHeapType) -> Self {
+        //TODO: SAFYTY
```
Comment (fanninpm) [inline vm/src/builtins/type.rs:45]: minor nitpick: spelling

```suggestion
        //TODO: SAFETY
```

**Comment by youknowone** [comment]
uuuh, i realized we cannot avoid lock due to our gil-less design. let me think about this.

File: vm/src/builtins/type.rs [hunk 2]
```diff
@@ -32,6 +34,28 @@ pub struct PyType {
...
+        unsafe { std::mem::transmute(heaptype) }
```
Comment (youknowone) [inline vm/src/builtins/type.rs:46]: This is not possible. I think this `From` impl never can exist. Their references can be casted, but not about the object body itself.

File: vm/src/object/core.rs [hunk 3]
```diff
@@ -110,7 +110,9 @@ struct PyInner<T> {
     typeid: TypeId,
     vtable: &'static PyObjVTable,
 
-    typ: PyRwLock<PyTypeRef>, // __class__ member
+    typ: PyTypeRef,
+    type_override: OnceCell<PyRwLock<PyTypeRef>>,
```
Comment (youknowone) [inline vm/src/object/core.rs:114]: I agree this is the only way we can practically accept for now, but growing `PyObject` body is not very pleasure.

File: vm/src/object/core.rs [hunk 4]
```diff
@@ -640,9 +643,29 @@ impl PyObject {
         }
     }
 
-    #[inline(always)]
-    pub(crate) fn class_lock(&self) -> &PyRwLock<PyTypeRef> {
-        &self.0.typ
+    pub(crate) fn get_class(&self) -> PyTypeRef {
```
Comment (youknowone) [inline vm/src/object/core.rs:646]:
```suggestion
    pub(crate) fn clone_class(&self) -> PyTypeRef {
```

**Pull request closed without merging**